### PR TITLE
v4.10.0

### PR DESCRIPTION
**Pull request by trevorgerhardt**
This should be tested on staging and requires a backend change (https://github.com/conveyal/analysis-backend/commit/1302e7ef2bcda918f07efcb7cce48636490b7778) which also needs to be tested on staging.

# Next Steps

Waiting for approval from either @abyrd or @ansoncfit. Approval will signify sufficient test procedures have been run on staging and with katalon and that this PR is ready to merge and be deployed to production.

## Testing priority
1. Transit mode selection in Add Trip Modifications and verifying they are turned on and off in a single point analysis.
2. Timetable dwell time and segment speed editing.

## Features
* Add transit mode selector for add trip pattern modifications.
* Adds a framework for creating and running database migrations.
* Show error when times are invalid.
* Split travel time display to have a total and moving time.
* Shows spinner on #473

## Bug Fixes
* Opportunities Upload / Download #887 
* Navigating to "Edit" from "Analysis" #890 
* Update travel times correctly on changes #886 
* Fix aggregation area upload bugs #889

### CODE REVIEW

File: lib/components/modification/__tests__/__snapshots__/reroute.js.snap [hunk 1]
```diff
@@ -1937,6 +1937,26 @@ exports[`Component > Modification > Reroute renders correctly 1`] = `
...
+    <h4
+      style={
+        Object {
+          "display": "flex",
+          "justifyContent": "space-between",
+        }
+      }
+    >
+      <span>
+        Total travel time
```
Comment (ansoncfit) [inline lib/components/modification/__tests__/__snapshots__/reroute.js.snap]: This makes sense for add-trip modifications, but could be confusing for reroute modifications, in which the displayed travel time is only for the added segment.  I'd suggest making this "Travel time" and adding the clarifications below.

File: lib/components/modification/__tests__/__snapshots__/reroute.js.snap [hunk 2]
```diff
@@ -1937,6 +1937,26 @@ exports[`Component > Modification > Reroute renders correctly 1`] = `
...
+      <span>
+        Total travel time 
+      </span>
+      <strong>
+        00:00:00
+      </strong>
+    </h4>
+    <p>
+      <em>
+        Dwell time at each stop + moving time along each segment
```
Comment (ansoncfit) [inline lib/components/modification/__tests__/__snapshots__/reroute.js.snap]: See above.  I'd suggest dwell time at added stops + moving time along added segments

File: lib/components/modification/segment-speeds.js [hunk 3]
```diff
@@ -160,51 +180,41 @@ export default class SegmentSpeeds extends PureComponent {
...
+
+    const averageSpeed = this._getAverageSpeed() || DEFAULT_SEGMENT_SPEED
+    const dwellSeconds = this._getDwellSeconds()
+    const movingSeconds = this._getMovingSeconds()
+    const totalSeconds = dwellSeconds + movingSeconds
 
     return (
       <>
+        <h4 style={{display: 'flex', justifyContent: 'space-between'}}>
+          <span>Total travel time </span>
```
Comment (ansoncfit) [inline lib/components/modification/segment-speeds.js]: This makes sense for add-trip modifications but could be confusing for reroute modifications, in which the displayed time is the total only for the rerouted portion.  I'd suggest making this "Travel time" and adding the clarifications below.

File: lib/components/modification/segment-speeds.js [hunk 4]
```diff
@@ -160,51 +180,41 @@ export default class SegmentSpeeds extends PureComponent {
...
+          <strong>{secondsToHhMmSsString(totalSeconds)}</strong>
+        </h4>
+        <p>
+          <em>Dwell time at each stop + moving time along each segment</em>
```
Comment (ansoncfit) [inline lib/components/modification/segment-speeds.js]: See above. I'd suggest "Dwell time at added stops + moving time along added segments"

File: lib/components/modification/segment-speeds.js [hunk 5]
```diff
@@ -217,14 +227,14 @@ export default class SegmentSpeeds extends PureComponent {
           value={toPrecision(averageSpeed)}
         />
         <MinutesSeconds
-          label={message('modification.addedSegments.time')}
+          label='Total moving time'
```
Comment (ansoncfit) [inline lib/components/modification/segment-speeds.js]: (along added segments)

File: lib/components/modification/segment-speeds.js [hunk 6]
```diff
@@ -139,11 +158,12 @@ export default class SegmentSpeeds extends PureComponent {
...
-        const travelTime = travelTimeSeconds - this._getTotalDwellSeconds()
         const totalKilometers = sum(segmentDistances)
         // figure out speed
         const speedKps =
-          travelTime >= 1 ? totalKilometers / travelTime : totalKilometers
+          travelTimeSeconds >= 1
+            ? totalKilometers / travelTimeSeconds
+            : totalKilometers
         const speedKph = speedKps * 3600
         update({segmentSpeeds: segmentDistances.map(() => speedKph)})
```
Comment (ansoncfit) [inline lib/components/modification/segment-speeds.js]: On staging, speeds (entered directly or calculated from an entered travel time) are rounded to the nearest km/hr when saving the modification.  Is that intentional?  More precision could be useful.

Comment (trevorgerhardt) [reply lib/components/modification/segment-speeds.js]: They're defined as `int`s in the backend model. Would need to update there to change.

**Comment by trevorgerhardt** [comment]
To resolve @ansoncfit's comments we are going to run a database migration converting old modifications to have a `transitMode` of `0`.

**Pull request merged**